src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -219,5 +221,64 @@ internal static Task<LightweightRenameLocations> FindRenameLocationsAsync(Soluti
             var renameLocations = await SymbolicRenameLocations.FindLocationsInCurrentProcessAsync(symbol, solution, options, cleanupOptions, cancellationToken).ConfigureAwait(false);
             return await ConflictResolver.ResolveSymbolicLocationConflictsInCurrentProcessAsync(renameLocations, newName, nonConflictSymbolKeys, cancellationToken).ConfigureAwait(false);
         }
+
+        /// <summary>
+        /// Creates a session between the host and OOP, effectively pinning this <paramref name="solution"/> until <see
+        /// cref="IDisposable.Dispose"/> is called on it.  By pinning the solution we ensure that all calls to OOP for
+        /// the same solution during the life of this rename session do not need to resync the solution.  Nor do they
+        /// then need to rebuild any compilations they've already built due to the solution going away and then coming
+        /// back.
+        /// </summary>
+        internal static IRemoteRenameKeepAliveSession CreateRemoteKeepAliveSession(
+            Solution solution,
+            IAsynchronousOperationListener listener)
+        {
+            return new RemoteRenameKeepAliveSession(solution, listener);
+        }
+
+        private sealed class RemoteRenameKeepAliveSession : IRemoteRenameKeepAliveSession
+        {
+            private readonly CancellationTokenSource _cancellationTokenSource = new();
+
+            public RemoteRenameKeepAliveSession(Solution solution, IAsynchronousOperationListener listener)
+            {
+                var cancellationToken = _cancellationTokenSource.Token;
+                var token = listener.BeginAsyncOperation(nameof(RemoteRenameKeepAliveSession));
+
+                var task = CreateClientAndKeepAliveAsync();
+                task.CompletesAsyncOperation(token);
+
+                return;
+
+                async Task CreateClientAndKeepAliveAsync()
+                {
+                    var client = await RemoteHostClient.TryGetClientAsync(solution.Services, cancellationToken).ConfigureAwait(false);
+                    if (client is null)
+                        return;
+
+                    // Now kick off the keep-alive work.  We don't wait on this as this will stick on the OOP side until
+                    // the cancellation token triggers.
+                    var unused = client.TryInvokeAsync<IRemoteRenamerService>(
+                        solution,
+                        (service, solutionInfo, cancellationToken) => service.KeepAliveAsync(solutionInfo, cancellationToken),
+                        cancellationToken).AsTask();
+                }
+            }
+
+            ~RemoteRenameKeepAliveSession()
+            {
+                if (Environment.HasShutdownStarted)
```

Comment:
wondering if this is necessary- there are multiple cases where the rename should have been disposed of that aren't a process shutdown.  For example solution close, file close, etc that would be more impactful if they failed.  It doesn't feel like this is helping that much?  Especially since integration tests don't shutdown the process

Reply:
so this is necessary as during shutdown finalizers may run, even if objects aren't disposed.  so we just don't do the finalizer check in that case (this is a standard pattern we have in code that validates that features are properly disposing objects for normal (non-shutdown) cases).